File: frontend/lib/api/geo.ts
import { http, qs } from "./http";

export interface GeoResult {
  label: string;
  display_name: string;
  lat: number;
  lng: number;
  country: string | null;
  country_code: string | null;
  region: string | null;
  city: string | null;
  place_type: string | null;
  osm_id: string | null;
}

export interface TargetArea {
  id: string;
  label: string;
  display_name: string | null;
  country: string | null;
  country_code: string | null;
  region: string | null;
  city: string | null;
  place_type: string | null;
  lat: number;
  lng: number;
  radius_km: number;
  mode: "include" | "exclude";
  workspace_id: string;
  created_at: string;
  updated_at: string;
}

export interface GeoSummary {
  areas: number;
  includes: number;
  excludes: number;
  countries: string[];
  estimated_reach: number;
  matched_contacts: number;
  provider: string;
}

export const geoApi = {
  search: (q: string) => http.get<{ results: GeoResult[] }>(`/geo/search${qs({ q })}`),
  reverse: (lat: number, lng: number) => http.get<GeoResult>(`/geo/reverse${qs({ lat, lng })}`),
  summary: () => http.get<GeoSummary>("/geo/summary"),
  listAreas: () => http.get<{ items: TargetArea[] }>("/geo/areas"),
  createArea: (body: Partial<GeoResult> & { lat: number; lng: number; radius_km: number; mode?: string; label: string }) =>
    http.post<TargetArea>("/geo/areas", body),
  updateArea: (id: string, body: { label?: string; radius_km?: number; mode?: string }) =>
    http.patch<TargetArea>(`/geo/areas/${id}`, body),
  deleteArea: (id: string) => http.del<void>(`/geo/areas/${id}`),
};

File: backend/app/schemas/geo.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class TargetMode(str, Enum):
    INCLUDE = "include"
    EXCLUDE = "exclude"


class GeoResult(BaseModel):
    label: str
    display_name: str
    lat: float
    lng: float
    country: Optional[str] = None
    country_code: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    place_type: Optional[str] = None
    osm_id: Optional[str] = None


class GeoSearchResponse(BaseModel):
    results: List[GeoResult]


class TargetAreaCreate(BaseModel):
    label: str = Field(..., min_length=1, max_length=200)
    display_name: Optional[str] = None
    country: Optional[str] = None
    country_code: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    place_type: Optional[str] = None
    lat: float
    lng: float
    radius_km: int = Field(25, ge=1, le=2000)
    mode: TargetMode = TargetMode.INCLUDE
    osm_id: Optional[str] = None


class TargetAreaUpdate(BaseModel):
    label: Optional[str] = None
    radius_km: Optional[int] = Field(None, ge=1, le=2000)
    mode: Optional[TargetMode] = None


class TargetAreaResponse(BaseModel):
    id: str
    label: str
    display_name: Optional[str] = None
    country: Optional[str] = None
    country_code: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    place_type: Optional[str] = None
    lat: float
    lng: float
    radius_km: int
    mode: TargetMode
    workspace_id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TargetAreaList(BaseModel):
    items: List[TargetAreaResponse]


class GeoSummary(BaseModel):
    areas: int
    includes: int
    excludes: int
    countries: List[str]
    estimated_reach: int
    matched_contacts: int
    provider: str
